### PR TITLE
Implemented single link instead of multiple links to read reviews

### LINKED ISSUE #8860: [a11y] Links to read reviews are confusing for screen reader/keyboard users
In the section where you can read reviews for each rating of an add-on (`.RatingsByStar-graph`), there are three links taking you to the same place, one for the star, one for the barContainer and one for the count. This results in the following output with the NVDA screen reader:

```
link  5
link  Read all five-star reviews
link  49
link  4
link  Read all four-star reviews
link  10
link  3
link  Read all three-star reviews
link  4
link  2
link  Read all two-star reviews
link  1
link  1
link  Read all one-star reviews
link  1
```

Aside from the extreme verbosity here, the meaning of the numbers isn't clear, especially when there are review counts between 1 and 5. Also, if you tab through this with the keyboard, you'll have to tab three times to get through the reviews for each rating, which provides no benefit because the links all go to the same place.

One way to solve this without changing the design too much is to hide two of the three links for each rating for accessibility (`aria-hidden="true"`) and give them `tabindex="-1"` so they aren't in the tab order but are still clickable. Note that if aria-hidden is used on an element, it is *critical* that that element is not in the tab order, as that will confuse users. The remaining (tabbable and not accessibility-hidden) link should get a title or label providing all of the information; e.g. "Read all 49 five-star reviews".

### PR DESCRIPTION
**Pull request by muskangupta-iitr**
This patch implements the approach of the single link instead of multi links as mentioned in the patch #9254 .
Fixes #8860 

@muffinresearch @willdurand @bobsilverberg Please review the code!

### CODE REVIEW

**Comment by willdurand** [comment]
How is this related to https://github.com/mozilla/addons-frontend/issues/8827?

**Comment by willdurand** [comment]
I am going to close this patch for now, even if that's just a draft. Please focus on:

- https://github.com/mozilla/addons-frontend/pull/9253
- https://github.com/mozilla/addons-frontend/pull/9254

**Comment by muskangupta-iitr** [comment]
@willdurand Sorry, by mistakenly I tagged different issue on this, I apologize for that. I updated the above comment. Really sorry for such unprofessional behavior. @muffinresearch Only told me to create this patch to try different approach for implementing single link instead of multiple links. Here is the [comment](https://github.com/mozilla/addons-frontend/pull/9254#issuecomment-604064434). I already successfully implemented the first approach as in the [comment](https://github.com/mozilla/addons-frontend/pull/9254#issuecomment-603899962)

**Comment by muskangupta-iitr** [comment]
@willdurand Can you review and open this patch please as I have to proceed further with these issues?

**Comment by muskangupta-iitr** [comment]
@muffinresearch @willdurand Should I create a new path for this?

**Comment by willdurand** [comment]
Oh, I see. Sorry, then let's reopen this PR and close thd other one, ok?

**Comment by muskangupta-iitr** [comment]
I had created two PRs according to two different approaches as discussed before. You can review the two PRs so that whichever is found suitable by you, could be further proceeded with while closing the other.

**Comment by muskangupta-iitr** [comment]
@bobsilverberg I tried the approach of making row into individual grid but in that case the width of each bar will be differently calculated and independent to each other that's why I breaked them into manual columns so that all the bar width will be calculated according to max-width of the column.

**Comment by bobsilverberg** [comment]
> @bobsilverberg I tried the approach of making row into individual grid but in that case the width of each bar will be differently calculated and independent to each other that's why I breaked them into manual columns so that all the bar width will be calculated according to max-width of the column.

Ah, I thought my method was working:

Screenshot 2020-03-31 17 18 41

but then realized (after you pointed it out) that if the review count was a different number of digits for different rows it doesn't work. Hmm, let me take another look.

**Comment by muskangupta-iitr** [comment]
@bobsilverberg Yes, I am also facing the same problem when the width of the third column will be different. Then, I switched to table for solving this.

**Comment by bobsilverberg** [comment]
@muskangupta-iitr But it looks like the same problem exists with your table approach:

Screenshot 2020-03-31 18 26 50

Are you not seeing this issue with your patch?

**Comment by muskangupta-iitr** [comment]
@bobsilverberg Here is the screenshot from my patch:

Screenshot from 2020-04-01 04-26-53

File: src/amo/components/RatingsByStar/index.js [hunk 1]
```diff
@@ -86,85 +90,104 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
     );
   }
 
+  handleRouting(star: string) {
+    const { addon, location, history, clientApp, siteLang } = this.props;
+    history.push({
+      pathname: `/${siteLang}/${clientApp}/addon/${addon.slug}/reviews/`,
+      search: `?score=${star}&src=${location.query.src}`,
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/index.js:97]: If `src` does not exist on the URL, then this yields a link with `src=undefined` in it, which I don't think we want.

File: src/amo/components/RatingsByStar/index.js [hunk 2]
```diff
@@ -86,85 +90,104 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
...
-
-    const linkTitles = {
-      /* eslint-disable quote-props */
-      '5': i18n.gettext('Read all five-star reviews'),
-      '4': i18n.gettext('Read all four-star reviews'),
-      '3': i18n.gettext('Read all three-star reviews'),
-      '2': i18n.gettext('Read all two-star reviews'),
-      '1': i18n.gettext('Read all one-star reviews'),
-      /* eslint-enable quote-props */
+    const self = this;
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/index.js:104]: Why is this needed?

File: src/amo/components/RatingsByStar/index.js [hunk 3]
```diff
@@ -86,85 +90,104 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
...
+            {['5', '4', '3', '2', '1'].map((star) => {
+              let starCount;
+              let starCountNode;
+
+              function createLink(body) {
+                invariant(addon, 'addon was unexpectedly empty');
+
+                return (
+                  <tr
+                    title={getLinkTitle(star, starCount) || ''}
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/index.js:137]: `getLinkTitle` returns `''` as the `default`, so why do you need the `|| ''` here?

File: src/amo/components/RatingsByStar/index.js [hunk 4]
```diff
@@ -29,11 +30,14 @@ type Props = {|
...
 type InternalProps = {|
   ...Props,
+  clientApp: string,
   dispatch: DispatchFunc,
   errorHandler: ErrorHandlerType,
   groupedRatings?: GroupedRatingsType,
+  history: ReactRouterHistoryType,
   i18n: I18nType,
   location: ReactRouterLocationType,
+  siteLang: ?string,
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/index.js:40]: I don't think we need the `?` here. `siteLang` should always exist.

File: src/amo/components/RatingsByStar/index.js [hunk 5]
```diff
@@ -86,85 +90,104 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
...
-                  )}
-                  <IconStar selected />
-                </div>
-                <div className="RatingsByStar-barContainer">
+              return (
+                <React.Fragment key={star}>
                   {loading ? (
-                    <div className="RatingsByStar-bar RatingsByStar-barFrame" />
+                    <tr>
+                      <th className="RatingsByStar-star">
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/index.js:158]: Why are you using `th` rather than `td`? As these are data rows, and now a header row, `td` seems more appropriate to me.

File: src/amo/components/RatingsByStar/index.js [hunk 6]
```diff
@@ -86,85 +90,104 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
...
-                    <LoadingText minWidth={95} />
-                  ) : (
-                    createLink(i18n.formatNumber(star))
-                  )}
-                  <IconStar selected />
-                </div>
-                <div className="RatingsByStar-barContainer">
+              return (
+                <React.Fragment key={star}>
                   {loading ? (
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/index.js:156]: I wonder if it would be a bit more efficient and cleaner if you were to produce the contents of the `<tr>` via a function, which returns either the loading version or the non-loading version, and then here either wrap that in `createLink` or not, depending on `loading`? This would allow you to avoid all the duplicate code for each `th` tag.

File: src/amo/components/RatingsByStar/index.js (hunk 1 above)
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:97]: Yes, we should add a conditional check for `src`. I will update that.

File: src/amo/components/RatingsByStar/index.js (hunk 4 above)
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:40]: Ohk! I will change this in the next update.

File: src/amo/components/RatingsByStar/index.js (hunk 5 above)
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:158]: Yes, I just tried the table approach with `th` tag, but `td` will be more appropriate. I will update that too.

File: src/amo/components/RatingsByStar/index.js (hunk 2 above)
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:104]: Beacuse I have to use `this` in `createLink` function. As, `this` will be undefined in the scope of that function.

File: src/amo/components/RatingsByStar/index.js (hunk 3 above)
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:137]: Yeah, I added this as extra check type but yes can be removed. I will updated that.

**Comment by muskangupta-iitr** [comment]
@bobsilverberg Done! I have updated the PR with the changes discussed above :+1:

File: src/amo/components/RatingsByStar/index.js [hunk 7]
```diff
@@ -86,85 +90,108 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
     );
   }
 
+  handleRouting(star: string) {
+    const { addon, location, history, clientApp, siteLang } = this.props;
+    history.push({
+      pathname: `/${siteLang}/${clientApp}/addon/${addon.slug}/reviews/`,
+      search: `?score=${star}&${
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/index.js:97]: I think you need to put the `&` inside the `{ }` for `location.query.src` because the way the code is now, without a `src` you still get a trailing `&` on the URL.

**Comment by muskangupta-iitr** [comment]
@bobsilverberg Umm, no actually I was trying few other approaches that could be used and was confused between which could be better so forgot to comment. I have already addressed the comment in my last commit with a different approach and would request you to review it and give suggestions so that we could implement whichever is best. :)

File: src/amo/components/RatingsByStar/index.js (hunk 7 above)
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:97]: Actually while writing tests for it I found a better approach for doing this. It goes like this:
```
    history.push(
      `/${siteLang}/${clientApp}/${reviewListURL({
        addonSlug: addon.slug,
        score: star,
        src: location.query.src,
      })}`,
    );
```
which will use the general function for creating url.

**Comment by muskangupta-iitr** [comment]
@bobsilverberg I have added the tests and changed the `history.push` path. So, I think this PR is ready for review?

File: src/amo/components/RatingsByStar/index.js [hunk 8]
```diff
@@ -86,85 +91,113 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
     );
   }
 
+  handleRouting(star: string) {
+    const { addon, location, history, clientApp, siteLang } = this.props;
+
+    invariant(addon, 'addon was unexpectedly empty');
+
+    history.push(
+      `/${siteLang}/${clientApp}${reviewListURL({
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/index.js:100]: This is better, nice job!

Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:100]: @bobsilverberg Thanks!

**Comment by muskangupta-iitr** [comment]
> Thanks for the updates @muskangupta-iitr. I added one comment, but also I don't think you addressed my comment at [#9295 (comment)](https://github.com/mozilla/addons-frontend/pull/9295#discussion_r401624501) yet.

@bobsilverberg Yes, the above changes looks really good. Thanks:). Should I replace mine with yours?

**Comment by bobsilverberg** [comment]
> > Thanks for the updates @muskangupta-iitr. I added one comment, but also I don't think you addressed my comment at [#9295 (comment)](https://github.com/mozilla/addons-frontend/pull/9295#discussion_r401624501) yet.
> 
> @bobsilverberg Yes, the above suggestion looks really better, could not think of this approach. Its actually a really good approach, reducing code redundancy as well as working really efficiently in my local system. Thanks for the suggestion. May I proceed with replacing my code with your suggestion?

Yes, please go ahead and use any or all of that suggestion in your patch.

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 9]
```diff
@@ -148,30 +155,59 @@ describe(__filename, () => {
...
+
+    function validateLink(row, score, expectedTitle) {
+      const clientApp = CLIENT_APP_ANDROID;
+      const lang = 'en-US';
+
+      expect(row).toHaveProp('onClick');
+
+      row.simulate('click', createFakeEvent({}));
+
+      sinon.assert.calledWithExactly(
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:169]: We tend to use just `sinon.assert.calledWith` as opposed to `sinon.assert.calledWithExactly`. Is there a reason you chose to use `calledWithExactly` here?

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 10]
```diff
@@ -148,30 +155,59 @@ describe(__filename, () => {
...
-    const bars = root.find('.RatingsByStar-barContainer').find(Link);
 
-    function validateLink(link, score, expectedTitle) {
-      expect(link).toHaveProp(
-        'to',
-        reviewListURL({ addonSlug: addon.slug, score }),
+    const tableRow = root.find('.RatingByStar-table-row');
+
+    function validateLink(row, score, expectedTitle) {
+      const clientApp = CLIENT_APP_ANDROID;
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:162]: This test is using values that will be present because they are the default values, but we really shouldn't rely on those in a test. If we're going to have an assertion that includes `clientApp` and/or `lang`, then we should explicitly set those in the test, and the assertion should check for the values we set. You can see an example of how this can be done in the test for ['renders an opensearch link for Firefox' in TestSearchForm.js](https://github.com/mozilla/addons-frontend/blob/master/tests/unit/amo/components/TestSearchForm.js#L179-L190).

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 11]
```diff
@@ -148,30 +155,59 @@ describe(__filename, () => {
...
+      const lang = 'en-US';
+
+      expect(row).toHaveProp('onClick');
+
+      row.simulate('click', createFakeEvent({}));
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:167]: You don't need the `{}` here, it can just be `createFakeEvent()`.

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 12]
```diff
@@ -148,30 +155,59 @@ describe(__filename, () => {
     const addon = addonForGrouping(grouping);
     store.dispatch(setGroupedRatings({ addonId: addon.id, grouping }));
     const root = render({ addon });
-    const counts = root.find('.RatingsByStar-star').find(Link);
-    const bars = root.find('.RatingsByStar-barContainer').find(Link);
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:152]: I think the description of this test should be updated, as I think it is no longer accurate. For example, it's not asserting anything about bars.

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 13]
```diff
@@ -185,22 +221,14 @@ describe(__filename, () => {
...
-
-    validateLink(counts.at(0), '5', 'Read all five-star reviews');
-    validateLink(counts.at(1), '4', 'Read all four-star reviews');
-    validateLink(counts.at(2), '3', 'Read all three-star reviews');
-    validateLink(counts.at(3), '2', 'Read all two-star reviews');
-    validateLink(counts.at(4), '1', 'Read all one-star reviews');
+    const tableRow = root.find('.RatingByStar-table-row');
+    const counts = tableRow.find('.RatingsByStar-count');
+
+    expect(counts.at(0).children()).toHaveText(grouping['5'].toString());
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:227]: I wonder if you can just add these assertions into the test you changed above, as opposed to having a separate test for it. It seems like it's just testing one more thing that appears on a row. The existing test is also checking the links, but we no longer have links on each of these.

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 14]
```diff
@@ -218,25 +246,38 @@ describe(__filename, () => {
...
       location: createFakeLocation({ query: { src } }),
     });
-    const counts = root.find('.RatingsByStar-count').find(Link);
+    const tableRow = root.find('.RatingByStar-table-row');
 
-    function validateLink(link, score) {
-      expect(link).toHaveProp(
-        'to',
-        reviewListURL({ addonSlug: addon.slug, score, src }),
+    const clientApp = CLIENT_APP_ANDROID;
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:251]: See my comment above about setting these in the test.

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 15]
```diff
@@ -218,25 +246,38 @@ describe(__filename, () => {
...
+    const lang = 'en-US';
+
+    function validateLink(row, score) {
+      expect(row).toHaveProp('onClick');
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:255]: I don't think we need this assertion. We already have one test that checks that the `onClick` prop exists when expected, and the next assertion will fail if the `onClick` prop doesn't exist.

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 16]
```diff
@@ -218,25 +246,38 @@ describe(__filename, () => {
...
+    validateLink(tableRow.at(0), '5');
+    validateLink(tableRow.at(1), '4');
+    validateLink(tableRow.at(2), '3');
+    validateLink(tableRow.at(3), '2');
+    validateLink(tableRow.at(4), '1');
   });
 
   it('renders IconStar', () => {
     const root = render();
-    const star = root.find(IconStar).at(0);
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:239]: Won't this still work? Are the changes to this test required?

File: src/amo/components/RatingsByStar/index.js [hunk 17]
```diff
@@ -86,85 +91,100 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
...
+        if (groupedRatings) {
+          starCount = groupedRatings[star];
+        }
+
+        starCountNode = i18n.formatNumber(starCount || 0);
+      }
+
+      const rowProps =
+        loading || !addon
+          ? {}
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/index.js:144]: I'm not sure we have a test for this logic. I.e., a test that asserts that `onClick` and `title` do not exist on the row when either `loading` is `true` or when we have no `addon`.

File: src/amo/components/RatingsByStar/index.js [hunk 18]
```diff
@@ -86,85 +91,100 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
...
+          <td className="RatingsByStar-star">
+            {!loading ? i18n.formatNumber(star) : <LoadingText width={100} />}
+            <IconStar selected />
+          </td>
+          <td className="RatingsByStar-barContainer">
+            {loading ? (
+              <div className="RatingsByStar-bar RatingsByStar-barFrame" />
+            ) : (
+              <div className="RatingsByStar-bar RatingsByStar-barFrame">
+                {starCount !== undefined
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/index.js:161]: Can you please check that we have a test for the case when `starCount` is `undefined`.

File: src/amo/components/RatingsByStar/styles.scss [hunk 19]
```diff
@@ -1,24 +1,44 @@
 @import '~amo/css/styles';
 
 .RatingsByStar-graph {
+  border-spacing: 0 6px;
   color: $grey-50;
-  display: grid;
+  // display: grid;
```
[the remark this answers is not in the document]
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/styles.scss:6]: Ohk! I will update that.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 9 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:169]: Like there is no specific reason for using it, but I saw some test in which `calledWithExactly` is used, in `TestSearchFilter`. i will update that.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 10 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:162]: Thanks @bobsilverberg :) Actually, I was looking this type of approach only but could not able to find it. I will update this.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 11 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:167]: Yes,I will update this.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 14 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:251]: Yes! I updated this too.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 15 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:255]: I will remove this in the next update.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 12 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:152]: @bobsilverberg Should I change the description or add test for bars ?

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 13 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:227]: Yes, we can do that. I will update that.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 12 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:152]: I my opinion because we have separate test case for bars, we can change the description of the test. So, after merging star count test description will be `renders star labels, star counts and links`.

File: src/amo/components/RatingsByStar/index.js (hunk 17 above)
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:144]: I think we can check the `onClick` prop for row in test `renders a loading state without an add-on`. Is this correct?

File: src/amo/components/RatingsByStar/index.js (hunk 18 above)
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:161]: No, we don't have any test for this case. I will add this in the update.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 10 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:162]: @bobsilverberg I tried this approach but getting an error.
This is the code I implemented:
```
  it('adds a `src` query parameter to the review links when available in the location', () => {
    const src = 'some-src';
    const grouping = {
      5: 964,
      4: 821,
      3: 543,
      2: 22,
      1: 0,
    };
    const addon = addonForGrouping(grouping);
    const clientApp = CLIENT_APP_FIREFOX;
    const lang = 'fr';

    const customStore = dispatchClientMetadata({ clientApp, lang }).store;
    customStore.dispatch(setGroupedRatings({ addonId: addon.id, grouping }));

    const root = render({
      addon,
      location: createFakeLocation({ query: { src } }),
    });
    const tableRow = root.find('.RatingByStar-table-row');

    function validateLink(row, score) {
      row.simulate('click', createFakeEvent());

      sinon.assert.calledWith(
        fakeHistory.push,
        `/${lang}/${clientApp}${reviewListURL({
          addonSlug: addon.slug,
          score,
          src,
        })}`,
      );
    }

    validateLink(tableRow.at(0), '5');
    validateLink(tableRow.at(1), '4');
    validateLink(tableRow.at(2), '3');
    validateLink(tableRow.at(3), '2');
    validateLink(tableRow.at(4), '1');
  });
```

Here is the error:
```
 AssertError: expected stub to be called with arguments

      241 |       row.simulate('click', createFakeEvent());
      242 | 
    > 243 |       sinon.assert.calledWith(
          |                    ^
      244 |         fakeHistory.push,
      245 |         `/${lang}/${clientApp}${reviewListURL({
      246 |           addonSlug: addon.slug,

      at Object.fail (node_modules/sinon/lib/sinon/assert.js:107:21)
      at failAssertion (node_modules/sinon/lib/sinon/assert.js:66:16)
      at Object.calledWith (node_modules/sinon/lib/sinon/assert.js:92:13)
      at validateLink (tests/unit/amo/components/TestRatingsByStar.js:243:20)
      at Object.<anonymous> (tests/unit/amo/components/TestRatingsByStar.js:253:5)

```

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 16 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:239]: Yes, it is working fine without any changes. I will update that.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 10 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:162]: I did a really stupid mistake :sweat_smile: I solved it. I forgot to pass history and store to render method. Thanks :)

**Comment by muskangupta-iitr** [comment]
@bobsilverberg I have added a commit updating the changes you suggested :+1:

**Comment by muskangupta-iitr** [comment]
@bobsilverberg A gentle reminder in case you missed the comment :) I had completed the patch with all the suggested changes, could you please review it?

**Comment by bobsilverberg** [comment]
My apologies @muskangupta-iitr. I was busy with some other things late last week, and then Friday was a holiday. I will take a look at this today.

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 20]
```diff
@@ -146,61 +153,66 @@ describe(__filename, () => {
...
+    const lang = 'en-US';
 
-    function validateLink(link, score, expectedTitle) {
-      expect(link).toHaveProp(
-        'to',
-        reviewListURL({ addonSlug: addon.slug, score }),
-      );
-      expect(link).toHaveProp('title', expectedTitle);
-    }
+    const customStore = dispatchClientMetadata({ clientApp, lang }).store;
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:159]: You don't need to create a new `customStore` here. `dispatchClientMetadata` accepts a `store` argument, so you could just change this to `dispatchClientMetadata({ clientApp, lang, store })` and then use `store.dispatch` in the next line.

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 21]
```diff
@@ -213,30 +225,44 @@ describe(__filename, () => {
       1: 0,
     };
     const addon = addonForGrouping(grouping);
-    store.dispatch(setGroupedRatings({ addonId: addon.id, grouping }));
+    const clientApp = CLIENT_APP_FIREFOX;
+    const lang = 'en-US';
+
+    const customStore = dispatchClientMetadata({ clientApp, lang }).store;
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:231]: As mentioned above, you don't need to create a new `store` to do this.

File: src/amo/components/RatingsByStar/index.js (hunk 17 above)
Comment (bobsilverberg) [reply src/amo/components/RatingsByStar/index.js:144]: Yes, we could add those assertions to that test, but it looks like you haven't done that in the last commit.

File: src/amo/components/RatingsByStar/index.js (hunk 18 above)
Comment (bobsilverberg) [reply src/amo/components/RatingsByStar/index.js:161]: I don't think that test is testing what I meant. If `starCount` is `undefined`, then rather than outputting the result of `this.renderBarValue(starCount)`, the code will output `null`, which would mean that there wouldn't even be a `div` with a `className` of `RatingsByStar-barValue`, but the test you mention (`renders 0% bar values when ratings do not exist`) is specifically checking elements with a `className` of `RatingsByStar-barValue`. 

So... either the code isn't working the way it is intended to, or we don't have a test where `starCount` is `undefined`.

**Comment by muskangupta-iitr** [comment]
> Thanks for the updates @muskangupta-iitr. I added some more comments about the tests.

Thanks @bobsilverberg for reviewing the updates. I will update the above changes in the next commit.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 12 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:152]: @bobsilverberg Thanks!

File: src/amo/components/RatingsByStar/index.js (hunk 17 above)
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:144]: Sorry, just forgot to add this one. I will update this in the next commit.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 20 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:159]: Thanks! I will update that in next update.

File: src/amo/components/RatingsByStar/index.js (hunk 18 above)
Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/index.js:161]: Yes, actually we don't have test when star count is undefined. I will add that in the next update.

**Comment by muskangupta-iitr** [comment]
@bobsilverberg I have updated the patch with the changes you suggested :+1:. Could you please review that?

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 22]
```diff
@@ -123,7 +130,9 @@ describe(__filename, () => {
 
   it('renders a loading state without an add-on', () => {
     const root = render({ addon: null });
+    const tableRow = root.find('.RatingByStar-table-row');
 
+    expect(tableRow.at(0)).not.toHaveProp('onClick');
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:135]: Thanks for adding this. We should also have an assert that checks that the prop `title` also does not exist.

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 23]
```diff
@@ -213,34 +227,65 @@ describe(__filename, () => {
...
     const star = root.find(IconStar).at(0);
+
     expect(star).toHaveLength(1);
     expect(star).toHaveProp('selected');
   });
 
+  it('renders when star count is undefined', () => {
+    let fiveStarCount;
+    const grouping = {
+      5: fiveStarCount,
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:275]: Do you need to create this `fiveStarCount` variable, or could you just have `5: undefined,` here?

File: tests/unit/amo/components/TestRatingsByStar.js [hunk 24]
```diff
@@ -213,34 +227,65 @@ describe(__filename, () => {
...
+      3: 0,
+      2: 0,
+      1: 0,
+    };
+    const addon = addonForGrouping(grouping);
+    store.dispatch(setGroupedRatings({ addonId: addon.id, grouping }));
+    const root = render({ addon });
+    const bar = root.find('.RatingsByStar-bar');
+
+    expect(bar.at(0).children()).toHaveLength(0);
```
Comment (bobsilverberg) [inline tests/unit/amo/components/TestRatingsByStar.js:286]: It would be nice to also have an assert for one of the bars that is not `undefined` which shows that `children()` has a length of `1`, so we can see that this test is in fact testing what we mean to test.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 22 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:135]: Ohk! I will add that.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 23 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:275]: Yes, we can directly write undefined but I just thought the case when `undefined` is possible, it happens only when variable is not initialized. But, yes it not necessary. I will update that.

File: tests/unit/amo/components/TestRatingsByStar.js (hunk 24 above)
Comment (muskangupta-iitr) [reply tests/unit/amo/components/TestRatingsByStar.js:286]: Ohk! Nice, I will add this too!

**Comment by muskangupta-iitr** [comment]
@bobsilverberg I have updated the changes and also did some styling changes as per the current design. Here are the screenshots:

Screenshot from 2020-04-15 02-43-02

Screenshot from 2020-04-15 02-43-06

File: src/amo/components/RatingsByStar/styles.scss [hunk 25]
```diff
@@ -2,23 +2,37 @@
...
 .RatingsByStar-graph {
   color: $grey-50;
-  display: grid;
   font-size: $font-size-default;
-  grid-gap: 12px;
-  grid-template-columns: max-content auto max-content;
-  // This controls whitespace and should also match the bar height.
   line-height: 16px;
+  width: 100%;
+  border-spacing: 0 8px;
```
Comment (bobsilverberg) [inline src/amo/components/RatingsByStar/styles.scss:8]: This is going to fail our linter. Properties need to be specified in alphabetical order. You can also see this in the Travis results.

Comment (muskangupta-iitr) [reply src/amo/components/RatingsByStar/styles.scss:8]: Yes, Thanks for that @bobsilverberg. I noticed it now only. I will update that.

**Comment by muskangupta-iitr** [comment]
> Looks great, thanks for all of your time and effort on this patch @muskangupta-iitr! There's just one small problem with the scss file that is causing the build to fail. Once that's fixed this should be good to land!

Thanks a lot @bobsilverberg for helping me completing this patch. I learned many things from this patch.

**Comment by muskangupta-iitr** [comment]
@bobsilverberg Done!

**Comment by muskangupta-iitr** [comment]
@bobsilverberg I have updated the title and description. Thanks for helping me out with the patch. Really glad to be able to contribute significantly. :)
